operator/cmd/k8s_identity.go

```diff
@@ -10,78 +10,208 @@ import (
 
 	"github.com/sirupsen/logrus"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/tools/cache"
 
 	"github.com/cilium/cilium/operator/identity"
 	"github.com/cilium/cilium/operator/metrics"
 	operatorOption "github.com/cilium/cilium/operator/option"
 	"github.com/cilium/cilium/operator/watchers"
 	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
 	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
 	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
 	"github.com/cilium/cilium/pkg/k8s/identitybackend"
-	"github.com/cilium/cilium/pkg/k8s/informer"
-	"github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/k8s/resource"
 	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/rate"
 )
 
-var identityStore cache.Store
+// IdentitiesGCCell is a cell that implements a periodic Cilium identities
+// garbage collector.
+// The GC subscribes to identities events to mark all the related identities
+// as alive. If an identity has no activity for a prolonged interval,
+// it is first marked for deletion and eventually deleted.
+// It is enabled only if the identities allocation mode is set to "CRD mode".
+var IdentitiesGCCell = cell.Module(
+	"k8s-identities-gc",
+	"Cilium identities CRD mode garbage collector",
 
-// deleteIdentity deletes an identity. It includes the resource version and
-// will error if the object has since been changed.
-func deleteIdentity(ctx context.Context, clientset k8sClient.Clientset, identity *v2.CiliumIdentity) error {
-	// Wait until we can delete an identity
-	err := identityRateLimiter.Wait(ctx)
-	if err != nil {
-		return err
+	cell.Provide(newIdentitiesGCConfig),
+	cell.Provide(newIdentitiesGC),
+
+	// Invoke forces the instantiation of the identity gc
+	cell.Invoke(func(*IdentitiesGC) {}),
+)
+
+// IdentitiesGCConfig contains the configuration for the identity-gc.
+type IdentitiesGCConfig struct {
+	AllocationMode string
+
+	IdentityGCInterval       time.Duration
+	IdentityHeartbeatTimeout time.Duration
+	EndpointGCInterval       time.Duration
+
+	RateInterval time.Duration
+	RateLimit    int64
+
+	MetricsEnabled bool
+}
+
+func newIdentitiesGCConfig(daemonCfg *option.DaemonConfig, operatorCfg *operatorOption.OperatorConfig) IdentitiesGCConfig {
+	return IdentitiesGCConfig{
+		AllocationMode:           daemonCfg.IdentityAllocationMode,
```

Comment:
What is `DaemonConfig` doing here? Do we have lots of places in the operator that pull that in? Worth perhaps making a list and fixing them soon.

Reply:
Solved using a separate structure to avoid pollution with global config options. I'm gonna make a list of such config options "cross-references".